src/wallet/rpcwallet.cpp

```diff
@@ -598,6 +653,103 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
         entry.pushKV("source", "legacy_hdseed");
 
         bool hasData = add_sapling(saplingAddresses, PaymentAddressSource::LegacyHDSeed, entry);
+        if (hasData) {
+            ret.push_back(entry);
+        }
+    }
+
+    // mnemonic seed source
+    {
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "mnemonic_seed");
+
+        bool hasData = false;
+        // transparent
+        // For each address in pwalletMain->mapKeyMetadata, include any address
+        // for which its HD keypath metadata indicates it is associated with the legacy
+        // account ID. All such addresses are derived from the mnemonic seed.
+        UniValue mnemonic_taddrs(UniValue::VARR);
+        for (const auto& [keyId, keyMeta] : pwalletMain->mapKeyMetadata) {
+            auto account = libzcash::ParseHDKeypathAccount(44, Params().BIP44CoinType(), keyMeta.hdKeypath);
+            if (account.has_value() && account.value() == ZCASH_LEGACY_ACCOUNT) {
+                mnemonic_taddrs.push_back(keyIO.EncodeDestination(keyId));
+                hasData = true;
+            }
+        }
+        if (hasData) {
+            // This extra "level" of JSON allows attributes to be added in
+            // the future in a backward-compatible way.
+            UniValue mnemonic_transparent(UniValue::VOBJ);
+            mnemonic_transparent.pushKV("addresses", mnemonic_taddrs);
+            entry.pushKV("transparent", mnemonic_transparent);
+        }
```

Comment:
Non-blocking: I think this works fine, but it would be nice to follow the pattern we have in the `legacy_random` source for consistency.